protocol-units/settlement/mcr/client/src/eth_stub.rs

```diff
@@ -0,0 +1,247 @@
+	EventNotificationStreamClosed,
+}
+
+// Codegen from artifact.
+sol!(
+	#[allow(missing_docs)]
+	#[sol(rpc)]
+	MCR,
+	"abi/mcr.json"
+);
+
+const MRC_CONTRACT_ADDRESS: &str = "0xBf7c7AE15E23B2E19C7a1e3c36e245A71500e181";
+const MAX_TX_SEND_RETRY: usize = 3;
+
+pub struct McrEthSettlementClient<P: Provider<T, Ethereum>, T: Transport + Clone> {
+	rpc_provider: P,
+	ws_provider: RootProvider<PubSubFrontend>,
+	gas_limit: u128,
+	_marker: PhantomData<T>,
+}
+
+impl<P: Provider<T, Ethereum>, T: Transport + Clone> McrEthSettlementClient<P, T> {
+	pub async fn build_with_provider<S>(
+		rpc_provider: P,
+		ws_url: S,
+		gas_limit: u128,
+	) -> Result<Self, anyhow::Error>
+	where
+		S: Into<String>,
+	{
+		let ws = WsConnect::new(ws_url);
+
+		let ws_provider = ProviderBuilder::new().on_ws(ws).await?;
+
+		Ok(McrEthSettlementClient {
+			rpc_provider,
+			ws_provider,
+			gas_limit,
+			_marker: Default::default(),
+		})
+	}
+}
+
+#[async_trait::async_trait]
+impl<P: Provider<T, Ethereum>, T: Transport + Clone> McrSettlementClientOperations
+	for McrEthSettlementClient<P, T>
+{
+	async fn post_block_commitment(
+		&self,
+		block_commitment: BlockCommitment,
+	) -> Result<(), anyhow::Error> {
+		let contract = MCR::new(MRC_CONTRACT_ADDRESS.parse().unwrap(), &self.rpc_provider);
+		let call_builder = contract.createBlockCommitment(
+			U256::from(block_commitment.height),
+			alloy_primitives::FixedBytes(block_commitment.commitment.0[..32].try_into()?),
+			alloy_primitives::FixedBytes(block_commitment.block_id.0[..32].try_into()?),
+		);
+		let MCR::createBlockCommitmentReturn { _0: eth_block_commitment } =
+			call_builder.call().await?;
+		let base_call_builder = contract.submitBlockCommitment(eth_block_commitment);
+
+		//validate gaz price
+		let mut estimate_gas = call_builder.estimate_gas().await?;
+		let gas_price = call_builder.provider.get_gas_price().await?;
+		let tx_fee_wei = estimate_gas * gas_price;
+
+		println!("estimate_gas:{estimate_gas} gas_price:{gas_price} tx_fee_wei:{tx_fee_wei}");
+
+		if tx_fee_wei > self.gas_limit {
+			return Err(McrEthConnectorError::GasLimitExceed(tx_fee_wei, self.gas_limit).into());
+		}
+
+		// Sending Tx automatically can lead to erros that depend on the state for Eth.
+		// It's convenient to manage some of them automatically to avoid to fail commitment Tx.
+		// I define a first one but other should be added depending on the test with mainnet.
+		for _ in 0..MAX_TX_SEND_RETRY {
+			let call_builder = base_call_builder.clone().gas(estimate_gas);
+			//send the Tx and wait for 2 confirmation.
+			let pending_tx = match call_builder.send().await {
+				Err(alloy_contract::Error::TransportError(TransportError::ErrorResp(payload))) => {
+					match payload.code {
+						//transaction underpriced
+						-32000 => {
+							if payload.message.contains("transaction underpriced") {
+								//increase gas of 10% and retry
+								estimate_gas += (estimate_gas * 10) / 100;
```

Comment:
It would be nice to extract this out into a strategy.

Reply:
OK I try to do some refactoring and define some rules and associated processing.

Reply:
I've created a simple rule based system to process send tx errors (see in send_eth_tx.rs).